cdc: add tracing

piodul opened the pull request:
Adds information about the stages of CDC mutation augmentation to tracing sessions.

Fixes https://github.com/scylladb/scylla/issues/5912
Tests: unit(debug, dev)

kbr- commented on cdc/log.cc:
```diff
@@ -989,10 +992,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
 
-            return f.then([trans = std::move(trans), &mutations, idx] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
+            return f.then([trans = std::move(trans), &mutations, idx, tr_state = std::move(tr_state)] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
+                tracing::trace(tr_state, "CDC: Generating log mutation [mutation #{}]", idx);
```
How are these indexes of any use?
Maybe we should show the pk, or token, or even both.

piodul commented:
@kbr-

Although I don't really use tracing during development, I can see its usefulness - it shows you the internal flow of one request, without the need to sift through logs on all participating replicas. Tracing can also be used in debugging performance issues in production - it can be configured to be enabled for a small percentage of all requests. Then, all recorded traces can be read from tables in `system_tracing` keyspace. That's at least what I inferred from our [tracing doc](https://github.com/scylladb/scylla/blob/master/docs/tracing.md) and Cassandra documentation on the net (@vladzcloudius please correct me if I'm wrong).

Could you tell me what additional information you would like to see included in the trace?

@tzach 

As the author of the issue, could you tell what kind of messages you would expect to be recorded in trace related to CDC? Are there any more requirements apart from improving debuggability?

denesb commented:
> I don't really see the point of this change.
> What is the trace for? What information should it give? Should the information be used for debugging, or performance analysis, or what? These questions should be answered before adding tracing functionality.
> 
> Currently it shows indexes in the `mutations` vector of mutations which are augmented. I don't really see the point of this.

Tracing is a very important tool that is used to debug misbehaving queries in customer clusters in a non-intrusive way. It is vastly superior to normal logging because it doesn't affect the node's overall performance (lower log levels often produce so much output that they make the node unusable), also they are enabled in the context of a query and can be enabled probabilistically or specifically for slow queries.

For this reason it is important that all aspects of query processing that have an impact on its latency/correctness are presented in the tracing log.

denesb commented on cdc/log.cc:
```diff
@@ -989,24 +992,34 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
                 // Note: further improvement here would be to coalesce the pre-image selects into one
                 // iff a batch contains several modifications to the same table. Otoh, batch is rare(?)
                 // so this is premature.
+                tracing::trace(tr_state, "CDC: Selecting preimage [mutation #{}]", idx);
```
There should be a message in the other branch as well. A missing message is strictly worse than a negative message, there is always a chance that a message is missing due to a bug and this creates uncertainty.

vladzcloudius replied:
Right. To sum it up: please, make sure to have at least one tracing message on each branch tree that you decided to trace.

piodul replied:
Added an `else` branch with an appropriate message.

piodul commented:
v2:

- Mutations in a batch are now differentiated in the trace using their `decorated_key`, not their index in the batch,
- Added a message about not selecting the preimage when it's not enabled,
- Fixed an issue about std::moving a lambda-captured, const variable.

piodul commented on cdc/log.cc:
```diff
@@ -989,24 +992,34 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
             }
 
-            return f.then([trans = std::move(trans), &mutations, idx] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
+            return f.then([trans = std::move(trans), &mutations, idx, tr_state = std::move(tr_state)] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
```
Good catch, I made the capturing lambda `mutable`.

slivne commented:
@piodul if you fixed all the comments ping reviewers

vladzcloudius commented on cdc/log.cc:
```diff
@@ -974,11 +976,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
-            [this, timeout, i] (std::vector<mutation>& mutations, service::query_state& qs) {
-        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            [this, timeout, i, tr_state = std::move(tr_state)] (std::vector<mutation>& mutations, service::query_state& qs) {
+        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs, tr_state] (int idx) mutable {
```
This won't work: "mutable" isn't going to cast away constantness. You need to instrument all lambdas in the call tree.

piodul replied:
Hmm, let's consider all `std::move(tr_state)`-s in this function (there are only two of them):

1. Capture at 983 moves out from the argument at 968 - OK, the argument isn't const,
1. Capture at 1005 moves out of capture at 984 - looks OK, the capturing lambda is mutable.

I tried verifying non-constness of the variables I was moving from by inserting `tr_state = nullptr` instructions one line before each of the `std::move`s, but the compilation failed.

It seems that when doing a simple capture by-copy (by placing only the name of the of the variable in the capture list, e.g. `[tr_state]`) in a mutable lambda, GCC does not remove constness from the type. I had to change the capture on the line above from `tr_state` to `tr_state = tr_state` to make `std::move`s from that variable work correctly.

Curiously, only GCC does that - in Clang and MSVC, variables captured that way in mutable lambdas seem to be non-const ([example on godbolt](https://godbolt.org/z/AWrzHK)).

vladzcloudius replied:
> Hmm, let's consider all `std::move(tr_state)`-s in this function (there are only two of them):
> 
>     1. Capture at 983 moves out from the argument at 968 - OK, the argument isn't const,
> 
>     2. Capture at 1005 moves out of capture at 984 - looks OK, the capturing lambda is mutable.

In order for tr_state to be not const you needed to make the lambda at 983 `mutable` .

I feel like there is some misunderstanding. Let me try to explain this in an example:

### Example 1 (bad)
```
some_movable_class var = ...;

// var is not const here
[var = std::move(var) ] {
    // var is const here
    [var = std::move(var)] () mutable { // var is going to be copied!
    }
}
```

### Example 2 (good)
```
some_movable_class var = ...;

// var is not const here
[var = std::move(var) ] () mutable{ // note the 'mutable' qualifier
    // var is not (!!!) const here
    [var = std::move(var)] () mutable { // var is moved
    }
}
```

`std::move()` is about r-values in general, where 'r' comes from "right".
It's important to make sure that the right operand in the assignment is not const.

> 
> 
> I tried verifying non-constness of the variables I was moving from by inserting `tr_state = nullptr` instructions one line before each of the `std::move`s, but the compilation failed.
> 
> It seems that when doing a simple capture by-copy (by placing only the name of the of the variable in the capture list, e.g. `[tr_state]`) in a mutable lambda, GCC does not remove constness from the type. I had to change the capture on the line above from `tr_state` to `tr_state = tr_state` to make `std::move`s from that variable work correctly.
> 
> Curiously, only GCC does that - in Clang and MSVC, variables captured that way in mutable lambdas seem to be non-const ([example on godbolt](https://godbolt.org/z/DkfkuR)).

piodul commented:
Update:

- added the `tr_state = tr_state` in capture (line 984)
- removed one unnecessary `mutable` specifier (line 1005)

denesb commented on cdc/log.cc:
```diff
@@ -993,24 +996,36 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
                 // Note: further improvement here would be to coalesce the pre-image selects into one
                 // iff a batch contains several modifications to the same table. Otoh, batch is rare(?)
                 // so this is premature.
+                tracing::trace(tr_state, "CDC: Selecting preimage, {}", m.decorated_key());
```
s/Selecting preimage, {}/Selecting preimage for {}/

There are similar messages below. It is confusing to see the key after the comma.

piodul replied:
Done, for other similar messages as well

denesb commented on cdc/log.cc:
```diff
@@ -993,24 +996,36 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
+                tracing::trace(tr_state, "CDC: Selecting preimage for {}", m.decorated_key());
                 f = trans.pre_image_select(qs.get_client_state(), db::consistency_level::LOCAL_QUORUM, m);
+            } else {
+                tracing::trace(tr_state, "CDC: Preimage not enabled for the table, so not selecting for {}", m.decorated_key());
```
CDC: Preimage not enabled for the table, not querying current value of {}

denesb commented on cdc/log.cc:
```diff
@@ -993,24 +996,36 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
+                tracing::trace(tr_state, "CDC: Generating log mutations for {}", m.decorated_key());
+                int generated_count;
                 if (should_split(m, *s)) {
+                    tracing::trace(tr_state, "CDC: Splitting into several log mutations for {}", m.decorated_key());
```
CDC: Splitting {} into several log mutations

piodul replied:
How about "CDC: Splitting {}"? Considering that trace messages are aligned to the right in cqlsh, keeping the key at the end of the message will make it easier to visually group messages for a particular key.

denesb replied:
ok.

denesb commented on cdc/log.cc:
```diff
@@ -993,24 +996,36 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
+                        ++generated_count;
                     });
                 } else {
+                    tracing::trace(tr_state, "CDC: No need to perform splitting for {}", m.decorated_key());
```
CDC: No need to split {}

denesb commented on cdc/log.cc:
```diff
@@ -993,24 +996,36 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
                     mutations.push_back(trans.transform(m, rs.get(), ts, tuuid, batch_no));
+                    generated_count = 1;
                 }
+                tracing::trace(tr_state, "CDC: Generated {} log mutations for {}", generated_count, m.decorated_key());
```
CDC: Generated {} log mutations from {}

piodul commented:
Update: applied @denesb's suggestions about trace messages' formulation

psarna commented:
Looks good to me too: f7c493b68. Also, congratulations on having so many reviewers on a single patch pull request!

denesb commented:
Short patches get the most reviews.

psarna commented:
Dequeued due to dtest failure in debug mode. Rafael reported:
```
This introduced a failure in debug unit tests.

Running

$ build/debug/test/boost/cdc_test -t test_map_logging --report_level=no
 --logger=HRF,test_suite:XML,test_suite,/home/espindola/scylla/scylla/testlog/debug/xml/cdc_test.1.xunit.xml
 --catch_system_errors=no -- -c2 -m2G --overprovisioned
 --unsafe-bypass-fsync 1 --blocked-reactor-notify-ms 2000000 --collectd
 0


Fails with the attached log.

Cheers,
Rafael
```
details can be found on scylladb-dev mailing list in this thread: `[QUEUED scylla next] cdc: add tracing`

piodul commented:
The line 1025 was the culprit:

```c++
tracing::trace(tr_state, "CDC: Generated {} log mutations from {}", generated_count, m.decorated_key());
```

Variable `m` is a reference to an element inside the `mutations` vector. At this point it might be invalidated because of a push_back to `mutations`.

I changed `m.decorated_key()` to `mutations[idx].decorated_key()` on that line, and added a warning comment. I ran unit tests in debug mode and they passed.

psarna commented:
Requeued as b1e8170bf9f9c442320e22c6db35b73d920147d

haaawk commented:
It's already in master. Closing...

vladzcloudius commented on cdc/log.cc:
```diff
@@ -974,11 +976,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
     return do_with(std::move(mutations), service::query_state(service::client_state::for_internal_calls(), empty_service_permit()),
-            [this, timeout, i] (std::vector<mutation>& mutations, service::query_state& qs) {
-        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            [this, timeout, i, tr_state = std::move(tr_state)] (std::vector<mutation>& mutations, service::query_state& qs) {
```
Missing `mutable`. 
As a result `tr_trace` (together with all other captured by-value variable) is going to be `const` inside this lambda and all further `std::move()s` on it are going to be copies.

vladzcloudius commented on cdc/log.cc:
```diff
@@ -974,11 +976,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
-            [this, timeout, i] (std::vector<mutation>& mutations, service::query_state& qs) {
-        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            [this, timeout, i, tr_state = std::move(tr_state)] (std::vector<mutation>& mutations, service::query_state& qs) {
+        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs, tr_state = tr_state] (int idx) mutable {
```
this `tr_state=tr_state` makes no sense to me.
The problem you tried to describe earlier is simply because of the missing `mutable` in the external lambda `tr_state` is already const here.

piodul replied:
This assignment removes constness of `tr_state`. The variable accessible from the inside will be modifiable.

You can verify that `tr_state` is modifiable inside this lambda - If you place `tr_state = nullptr;` on the next line, it will work.

vladzcloudius replied:
It's not removing the constness from `tr_state` - it declares a variable with the name `tr_state` in the context of the internal lambda which initialized using a copy-constructor with a value of a `tr_state` from the external lambda.

The `tr_state` in the internal lambda is non-const because you used `mutable` in the declaration of the internal lambda.

This code is functionally correct (I thought it wasn't but now I see that it is) but it's very cryptic and hacky.

I suggest you either put a very fat comment that explains that variable copy trick, or you can put `mutable` in the external lambda declaration and thereby avoid the need for the "trick".

I'd personally prefer the later.

piodul replied:
> This code is functionally correct (I thought it wasn't but now I see that it is) but it's very cryptic and hacky.
> I suggest you either put a very fat comment that explains that variable copy trick, or you can put mutable in the external lambda declaration and thereby avoid the need for the "trick".

With the "trick" we have better const correctness (because we don't have to make every lambda on the path to this mutable), but I guess that if it comes as too surprising and cryptic, then the benefit is lost. I will change to `mutable` in a follow up PR.
